pkg/controller/master/rancher/pod_resources.go

```diff
@@ -0,0 +1,132 @@
+		node.Annotations[LimitsAnnotation] != limits {
+		toUpdate := node.DeepCopy()
+		if toUpdate.Annotations == nil {
+			toUpdate.Annotations = map[string]string{}
+		}
+		toUpdate.Annotations[RequestsAnnotation] = requests
+		toUpdate.Annotations[LimitsAnnotation] = limits
+		return h.nodeController.Update(toUpdate)
+	}
+
+	return node, nil
+}
+
+func (h *Handler) getNonTerminatedPods(node *corev1.Node) ([]*corev1.Pod, error) {
+	var pods []*corev1.Pod
+	fromCache, err := h.podCache.List("", labels.NewSelector())
+	if err != nil {
+		return pods, err
+	}
+
+	for _, pod := range fromCache {
+		if pod.Spec.NodeName != node.Name {
+			continue
+		}
+		// kubectl uses this cache to filter out the pods
+		if pod.Status.Phase == "Succeeded" || pod.Status.Phase == "Failed" {
+			continue
+		}
+		pods = append(pods, pod)
+	}
+	return pods, nil
+}
+
+func getPodResourceAnnotations(pods []*corev1.Pod) (string, string, error) {
+	requests, limits := aggregateRequestAndLimitsForNode(pods)
+	requestsBytes, err := json.Marshal(requests)
+	if err != nil {
+		return "", "", err
+	}
+
+	limitsBytes, err := json.Marshal(limits)
+	return string(requestsBytes), string(limitsBytes), err
+}
+
+func aggregateRequestAndLimitsForNode(pods []*corev1.Pod) (map[corev1.ResourceName]resource.Quantity, map[corev1.ResourceName]resource.Quantity) {
+	requests, limits := map[corev1.ResourceName]resource.Quantity{}, map[corev1.ResourceName]resource.Quantity{}
+	for _, pod := range pods {
+		podRequests, podLimits := getPodData(pod)
+		addMap(podRequests, requests)
+		addMap(podLimits, limits)
+	}
+	if pods != nil {
+		requests[corev1.ResourcePods] = *resource.NewQuantity(int64(len(pods)), resource.DecimalSI)
+	}
+	return requests, limits
+}
+
+func getPodData(pod *corev1.Pod) (map[corev1.ResourceName]resource.Quantity, map[corev1.ResourceName]resource.Quantity) {
+	requests, limits := map[corev1.ResourceName]resource.Quantity{}, map[corev1.ResourceName]resource.Quantity{}
+	for _, container := range pod.Spec.Containers {
+		addMap(container.Resources.Requests, requests)
+		addMap(container.Resources.Limits, limits)
+	}
+
+	for _, container := range pod.Spec.InitContainers {
+		addMapForInit(container.Resources.Requests, requests)
+		addMapForInit(container.Resources.Limits, limits)
+	}
+	return requests, limits
+}
+
+func addMap(data1 map[corev1.ResourceName]resource.Quantity, data2 map[corev1.ResourceName]resource.Quantity) {
+	for name, quantity := range data1 {
+		if value, ok := data2[name]; !ok {
+			data2[name] = quantity.DeepCopy()
+		} else {
+			value.Add(quantity)
+			data2[name] = value
+		}
+	}
+}
+
+func addMapForInit(data1 map[corev1.ResourceName]resource.Quantity, data2 map[corev1.ResourceName]resource.Quantity) {
+	for name, quantity := range data1 {
+		value, ok := data2[name]
+		if !ok {
+			data2[name] = quantity.DeepCopy()
+			continue
+		}
+		if quantity.Cmp(value) > 0 {
+			data2[name] = quantity.DeepCopy()
```

Comment:
Question: 
Why `addMapForInit` is different with `addMap` in processing the `quantity`?

Reply:
It may not be fully correct here. 

As `container` state is not checked, the last `init container` with  resource requests/limits value will overwrite others value, no matter which `init container` is trully occupying resources.

Normally, when the `initContainers` finished, they will release the resources, their related values seems should be 0 then.

Reply:
The calculation here should be the minimum resources needed to make sure the POD is running/restart properly

Reply:
Totally，the result here is `request`, not `used`